Sources/UID2/UID2Manager.swift

```diff
@@ -187,13 +232,84 @@ public final actor UID2Manager {
         self.identity = validIdentity
-        KeychainManager.shared.saveIdentityToKeychain(validIdentity)
+        KeychainManager.shared.saveIdentityToKeychain(validatedIdentityPackage)
+        
+        await checkIdentityRefresh()
+        await checkIdentityExpiration()
         
         return validIdentity
     }
 
     // MARK: - Refresh and Timer
     
+    private func checkIdentityRefresh() async {
+        refreshJob?.cancel()
+        refreshJob = nil
+
+        if !automaticRefreshEnabled {
+            return
+        }
+        
+        if let identity = identity {
+            // If the identity is already suitable for a refresh, we can do so immediately. Otherwise, we will work out
+            // how long it is until a refresh is required and schedule it accordingly.
+            if await hasExpired(expiry: identity.refreshFrom) {
+                refreshJob = Task {
+                    await refreshToken(identity: identity)
+                }
+            } else {
+                refreshJob = Task {
+                    let now = Date().millisecondsSince1970
+                    let refreshFrom = identity.refreshFrom
+                    let delayInMilliseconds = refreshFrom - now
+                    let delayInNanoseconds = UInt64(delayInMilliseconds * 1000000)
+                    try await Task.sleep(nanoseconds: delayInNanoseconds)
+                    await refreshToken(identity: identity)
+                }
+            }
+        }
+    }
+    
+    /// The identity status can change as we reach specific time events. We want to observe these and make sure that when
+    /// they are reached, we can report them accordingly to our consumer.
+    private func checkIdentityExpiration() async {
+        checkRefreshExpiresJob?.cancel()
+        checkRefreshExpiresJob = nil
+        
+        checkIdentityExpiresJob?.cancel()
+        checkIdentityExpiresJob = nil
+        
+        if !checkExpiration {
+            return
+        }
+        
+        if let identity = identity {
+            
+            // If the expiration time of being able to refresh is in the future, we will schedule a job to detect if we
+            // pass it. This will allow us to reevaluate our state and update accordingly.
+            if await !hasExpired(expiry: identity.refreshExpires) {
+                checkRefreshExpiresJob = Task {
+                    let now = Date().millisecondsSince1970
+                    let diffToNow = identity.refreshExpires - now
+                    let delayInNanoseconds = UInt64(diffToNow * 1000000)
+                    try await Task.sleep(nanoseconds: delayInNanoseconds)
+                    await validateAndSetIdentity(identity: identity, status: nil, statusText: nil)
+                }
+            }
+            
+            if await !hasExpired(expiry: identity.identityExpires) {
+                checkIdentityExpiresJob = Task {
+                    let now = Date().millisecondsSince1970
```

Comment:
These three lines are repeated three times in the code, worth a separate function?

```
                     let now = Date().millisecondsSince1970
                     let diffToNow = identity.refreshExpires - now
                     let delayInNanoseconds = UInt64(diffToNow * 1000000)
```

Reply:
Done.